build.jsh

```diff
@@ -0,0 +1,60 @@
+//usr/bin/env jshell --show-version --execution local "$0" "$@"; exit $?
+
+/*
+ * Download "Bach.java" and "Bach.jsh" from github to local "target" directory.
+ */
+Path target = Files.createDirectories(Paths.get("target"))
+URL context = new URL("https://raw.githubusercontent.com/sormuras/bach/master/src/bach/")
+for (Path script : Set.of(target.resolve("Bach.java"), target.resolve("Bach.jsh"))) {
+    if (Files.exists(script)) continue; // comment to force download files
+    try (InputStream stream = new URL(context, script.getFileName().toString()).openStream()) {
+        Files.copy(stream, script, StandardCopyOption.REPLACE_EXISTING);
+    }
+}
+
+/*
+ * Switch Bach to verbose mode. That'll print the commands before execution.
+ */
+System.setProperty("bach.verbose", "true")
+
+/*
+ * Source "Bach.java" and "Bach.jsh" into this jshell session.
+ */
+/open target/Bach.java
+/open target/Bach.jsh
+
+/*
+ * Define global properties.
+ */
+String name = "org.codefx.demo.jpms_hello_world"
+String main = "org.codefx.demo.jpms.HelloModularWorld"
+Path classes = Paths.get("target/bach/classes")
+Path sources = Paths.get("src/main/java")
+Path archive = Paths.get("target/bach/jpms-hello-world.jar")
+
+/*
+ * Compile.
+ *
+ * Compile using "--module-source-path" and "--module" combo fails:
+ * <pre>
+ *   javac("-d", "target/bach/classes", "--module-source-path", "src/main/java", "--module", name)
+ *   error: module org.codefx.demo.jpms_hello_world not found in module source path
+ * </pre>
+ * So, we have to simulate "$(find src -name '*.java')" and use "Bach.Command" directly.
+ */
+Bach.Command javac = new Bach.Command("javac")
+javac.add("-d").add(classes)
+javac.addAll(sources, Bach.Basics::isJavaFile)
+javac.run()
+
+/*
+ * Package.
+ */
+jar("--create", "--file", archive, "--main-class", main, "-C", classes, ".")
+
+/*
+ * Launch.
+ */
+java("--module-path", archive, "--module", name)
```

Comment:
This part doesn't work on my machine:

```
running java with 4 argument(s)
java
--module-path
  target/bach/jpms-hello-world.jar
--module
  org.codefx.demo.jpms_hello_world
Unrecognized option: --module-path
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
|  java.lang.AssertionError thrown: expected an exit code of zero, but got: 1
|        at Bach$Command.run (#19:314)
|        at Bach.run (#19:12)
|        at java (#20:1)
|        at (#40:1)
```

The ones before work and I verified that the created JAR works as a module on Java 9.

Maybe it's my setup's fault? I don't have Java 9 "installed", I just unzip the archive, and symlink `java9`, `javac9`, ... to the 9 variants of those commands. Because there was no `jshell` in 8, that command works without `9` (which I assume is the reason why the script works at all. I tentatively set `$JAVA_HOME` to my JDK 9 folder but that didnÄt help either.

Ideas?

Reply:
Ja. "java" is not available via `ToolProvider`. Bach resorts to the command line and system path to execute "java" as an external tool. If you have the JRE-9 "java" first in your path, it should work.

Trying "JAVA_HOME/bin/java[.exe]" in Bach should help.

Reply:
> Trying "JAVA_HOME/bin/java[.exe]" in Bach should help.

Sorry for being stupid, but where would I put that?

Reply:
I'll update Bach in two minutes -- extracting the Java 9 binary from the jshell process. :)

Reply:
Nice.